Update the Linux OTA guide to take into account memfaultd (#270)
memfault/memfault-docs: docs/linux/linux-releases-integration-guide.mdx

### Create an OTA Payload

> 

Reviewer: Should we explain the different tab options somewhere at the beginning?
I feel like this is easily misunderstood.

Maybe we can add something along these lines to the Prerequisites:

### memfaultd (recommended)
We strongly recommend integrating memfaultd first (link to getting started). A key function of memfaultd is to correctly configure and control SWUpdate. While it is possible to use SWUpdate with Memfault's OTA backend without memfaultd, you will have to manage the custom configuration of SWUpdate yourself. Similarly, it's possible to use Memfault's OTA backend with a custom OTA client. We provide limited support for the use cases without memfaultd.

In the sections below, select the tab of the use case for which use would like to view detailed instructions:
- SWUpdate (with memfaultd) -- recommended
- SWUpdate (custom) -- not recommended
- Custom -- not recommended

Author: Yeah! I thought about this but then forgot. Will mostly use your suggestion!

<LinuxOtaOptionsTabs>
  <TabItem value="swupdate-memfaultd">

SWUpdate expects OTA Payloads in its own `.swu` format. Read [documentation on
SWUpdate][3] to learn how to create `.swu` files for your project. Our [Yocto
example][source-yocto-example] contains a working SWUpdate image

> `memfault-image`.

Reviewer: Maybe mentioned `bitbake memfault-image` and then were the .swu can be found in the build dir?

Not sure if/where we explain the various env vars that are used in the example (`MEMFAULT_SOFTWARE_VERSION`, `MEMFAULT_SOFTWARE_TYPE`, etc.).

Would it make sense to explain that here? If you're evaluating Memfault, it would be nice to just be able to have fully working steps that leverage the meta-memfault-example.

Or maybe better to add a README.md to the meta-memfault-example?

Author: :+1: I wanna add a README to the example dir https://memfault.slack.com/archives/C03P60SCLMN/p1658846875063979?thread_ts=1658834872.702339&cid=C03P60SCLMN

Author: Will do elsewhere. But I will add a stronger callout about the example dir here.

to download the Release on a device!

<LinuxOtaOptionsTabs>
  <TabItem value="swupdate-memfaultd">

Memfault supports [SWUpdate][3] as a way to update the software on an Embedded

> Linux device using its [Suricatta][4] daemon. It communicates with the server

Reviewer: Ambiguous what "its" is referring to
```suggestion
Linux device using [SWUpdate's Suricatta][4] daemon. It communicates with the server
```

using the [hawkBit][5] DDI API standard.

> 

Reviewer: I wonder if we should highlight/call out the example layer here more front-and-center as an example of how the following steps will look like once performed.

Author: https://github.com/memfault/memfault-docs/pull/270/commits/645cbf4bfa44e9cb007ba75374192cab7cdfade3

#### Building SWUpdate to include Suricatta

Find the following options, either via `menuconfig` or by modifying `defconfig`,
and add them to `defconfig` in a recipe that appends to `swupdate` (here's [an
example][source-swupdate-defconfig]):

- `CONFIG_SURICATTA=y`: configures SWUpdate to include the Suricatta daemon in
  the build.
- `CONFIG_SURICATTA_HAWKBIT=y`: configures Suricatta to use hawkBit server mode.
- Unset `CONFIG_SURICATTA_GENERAL`: SWUpdate works either in hawkBit mode or a
  general-purpose HTTP server mode. Disable general-purpose HTTP server support.

#### Configuring the Suricatta on-device daemon to use Memfault

The [`memfaultd`][source-memfaultd] daemon includes a SWUpdate plugin. You can
[control which plugins are built-in][docs-linux-control-plugins], but by
default, it builds with the SWUpdate plugin enabled.

The [SWUpdate plugin for `memfaultd`][source-memfaultd-swupdate-plugin]

> constructs a valid SWUpdate configuration in a temporary directory file, and

Reviewer: ```suggestion
constructs a valid SWUpdate configuration in a temporary file, and
```

> instructs SWUpdate to use it as its configuration file. However, to allow for

Reviewer: ```suggestion
which SWUpdate then uses it as its configuration file. However, to allow for
```
Technically speaking, memfaultd doesn't instruct swupdate. It's that 09-swupdate-args file that sets the config file path. Maybe link to that here?

Author: :+1: Will also need to link to that reference (to be added) about `/etc/memfaultd.conf`, where the path to `/tmp/swupdate.cfg` can be specified (I think).

Author: https://github.com/memfault/memfault-docs/pull/270/commits/5a11e5a76013f2a5331765ef29f98ef95397bcf6

(to be added: reference to `/etc/memfaultd.conf`)

> further configuration, it also includes `/etc/swupdate.cfg`. Feel free to

Reviewer: I would maybe flip this around:
The .. plugin writes a SWUpdate configuration in a temporary file, using `/etc/update.cfg` as (optional) base configuration, extending it with the pieces of configuration necessary for SWUpdate to work with Memfault's OTA servers.

Author: https://github.com/memfault/memfault-docs/pull/270/commits/10ad78af64c8c5d927e6c0a2699d16d85074e3d2

configure your SWUpdate installation using this file, e.g. by adding a recipe
that appends to `swupdate` (see [this example][source-swupdate-bbappends]).

Note that the `identify` and the `suricatta` sections are controlled by

> `memfaultd` and should not be specified in `/etc/swupdate.cfg`.

Reviewer: Maybe also note that the paths of both the input/base and output file paths can be overridden via memfaultd.conf?

Author: https://github.com/memfault/memfault-docs/pull/270/commits/f43c30bd5bb94b072a221c8aa4a4904786b7fbbd

#### SWUpdate plugin dependencies

The dependencies of the SWUpdate plugin are `u-boot-env` and `u-boot-fw-utils`.
Make sure to include those in your image file like so:

```
IMAGE_INSTALL:append = " memfault-device-info u-boot-env u-boot-fw-utils"
```

Note that `memfault-device-info` is a required executable for `memfaultd`. Read
[our docs on `memfault-device-info`][docs-memfault-device-info] for more
information.

#### Add `/etc/fw_env.config` to your build

See [this example `/etc/fw_env.config` file][source-fwenv-config]. SWUpdate as

> well as our our reboot reasons plugin depend on it.

Reviewer: ```suggestion
well as `memfaultd`'s reboot reason tracking plugin depend on it.
```
